### DISCUSSION TITLE
Metadata API: support custom property attribute values (Facebook sharing)

### DISCUSSION THREAD

**Discussion by ArturAmpilogov** [vercel/next.js#57527] [Ideas]
### Goals

1. Support  custom property values: `<meta property="MyCustomProperty" content="MyContent">`.
2. Allow to override the metadata in child layouts and components via `generateMetadata` or exporting static metadata. Currently removing a metadata tag in child components is not possible for tags declared in the `head`:

```
// (root) layout.tsx
<head>
   <meta property="fb:app_id" content="12345" />
</head>
```

### Background

1. Facebook uses `og` properties for metadata, except the `fb:app_id`, which is [required](https://developers.facebook.com/docs/sharing/opengraph/using-objects/).
Example: `<meta property="fb:app_id" content="1234567890" /> `
2. Current `Metadata` type in Next.js supports `other` fields:
```
other?: {
    [name: string]: string | number | Array<string | number>;
  };
 ```
 
The problem is that it will resolve to the `<meta name=...`, while custom `<meta property=...` is not supported.

### Proposal

1. It seems there are not so many tags missed, like `meta http-equiv="..."`, and the code in Next.JS already picked up the path to include mostly used meta attributes (`publisher`, `manifest`, etc). Thus is seems there is no need to provide arbitrary attribute creation like `<meta my-attr="...`.
 
2. Introduce new types

    ```
    type MetadataNames = [name: string]: string | number | Array<string | number> & DeprecatedMetadataFields;
    type MetadataProperties = [property: string]: string | number | Array<string | number>
    ```

3. Extend the `Metadata.other` field with the properties, also support the previous API:

    before:
    ```
    other?: { [name: string]: string | number | Array<string | number> } & DeprecatedMetadataFields;
    ```

    after:
    ```
    other?: MetadataNames | { names: MetadataNames; properties: MetadataProperties };
    ```

----

For Facebook sharing it will look like:
```
const meta: Metadata = {
  other: {
   properties: {
      "fb:app_id": "1234567890" 
   }
}
```

**Comment by mhaneef50673** [comment]
+1

**Comment by shadowprompt** [comment]
+1

**Comment by Dwlad90** [comment]
+1

**Comment by Nekvit** [comment]
+1

**Comment by kevinmmarlow** [comment]
It seems reasonable that we could allow for `other` to take a custom `resolver` function that returns a `Meta` type object. This would be more flexible than `properties` as it allows for full flexibility to construction of the metadata and can more easily expand to other types in the future if the `Meta` interface changes.

**Comment by zhangyu1818** [comment]
I think it is possible to add an additional array option for the `other` field so that it does not affect the previous functionality.


```typescript
export const metadata: Metadata = {
  other: [
    { property: "product:price:amount", content: "9.99"}
  ]
};
```

**Reply by zhangyu1818** [reply]
I created a PR https://github.com/vercel/next.js/pull/66996.

**Comment by revskill10** [comment]
+2

**Comment by denisemenov** [comment]
+1

Another option - add 'fb' property to OpenGraph section.

```
const metadata: Metadata = {
  openGraph: {
   fbAppId:  "1234567890" 
  }   
}
```

**Reply by Netail** [reply]
I think
```
const metadata: Metadata = {
  facebook: {
   appId:  "1234567890" 
  }   
}
```
is more appropriate as it has a facebook prefix (`fb:`) instead of open graph (`og:`)

**Reply by hishamank** [reply]
yeah this too, but @zhangyu1818 solution would solve other requirements too

**Comment by mariasimo** [comment]
We also need this

**Comment by Henkedanne** [comment]
very much need this!

**Comment by brunom63** [comment]
+1

**Comment by Compuponder** [comment]
Per https://nextjs.org/docs/app/api-reference/functions/generate-metadata#facebook it should be working, but it does not.  Problem "Object literal may only specify known properties and 'facebook' does not exist in type 'Metadata'" (Code 2352) is thrown by TypeScript.

Next 14.2.5

**Reply by polesapart** [reply]
@Compuponder IIRC i'ts in next 15 canaries only. It seems the documentation leaked to stable version by accident.

**Comment by swax** [comment]
Facebook metadata property works in Next.js 15 and later

```
export const metadata = {
  facebook: {
    appId: '12345678',
  },
}
```
